recipes/cache/tree_cache.go

```diff
@@ -0,0 +1,298 @@
+		if !exists {
+			return nil, ErrNodeNotFound
+		}
+		current = next
+	}
+
+	return current, nil
+}
+
+// CurrentChildren returns the current set of children at the given full path, mapped by child name.
+// There are no guarantees of accuracy; this is merely the most recent view of the data.
+// If there is no node at this path, ErrNodeNotFound is returned.
+func (tc *TreeCache) CurrentChildren(fullPath string) (map[string]*ChildData, error) {
+	node, err := tc.findNode(fullPath)
+	if err != nil {
+		return nil, err
+	}
+	if node.state != NodeStateLIVE {
+		return nil, ErrNodeNotLive
+	}
+
+	m := make(map[string]*ChildData, len(node.children))
+	for child, childNode := range node.children {
+		// Double-check liveness after retreiving data.
+		if childNode.childData != nil && childNode.state == NodeStateLIVE {
+			m[child] = childNode.childData
+		}
+	}
+
+	// Double-check liveness after retreiving children.
+	if node.state != NodeStateLIVE {
+		return nil, ErrNodeNotLive
+	}
+	return m, nil
+}
+
+// CurrentData returns the current data for the given full path.
+// There are no guarantees of accuracy. This is merely the most recent view of the data.
+// If there is no node at the given path, ErrNodeNotFound is returned.
+func (tc *TreeCache) CurrentData(fullPath string) (*ChildData, error) {
+	node, err := tc.findNode(fullPath)
+	if err != nil {
+		return nil, err
+	}
+	if node.state != NodeStateLIVE {
+		return nil, ErrNodeNotLive
+	}
+
+	return node.childData, nil
+}
+
+// callListeners calls all listeners with given event.
+// Error is handled by handleException().
+func (tc *TreeCache) callListeners(evt TreeCacheEvent) {
+	tc.listeners.ForEach(func(listener interface{}) {
+		if err := listener.(TreeCacheListener).ChildEvent(tc.client, evt); err != nil {
+			tc.handleException(err)
+		}
+	})
+}
+
+// handleException sends an exception to any listeners, or else log the error if there are none.
+func (tc *TreeCache) handleException(e error) {
+	if tc.errorListeners.Len() == 0 {
+		tc.logger.Printf("%s", e)
+		return
+	}
+	tc.errorListeners.ForEach(func(listener interface{}) {
+		listener.(curator.UnhandledErrorListener).UnhandledError(e)
+	})
+}
+
+func (tc *TreeCache) handleStateChange(newState curator.ConnectionState) {
+	switch newState {
+	case curator.SUSPENDED:
+		tc.publishEvent(TreeCacheEventConnSuspended, nil)
+	case curator.LOST:
+		tc.isInitialized.UnSet()
+		tc.publishEvent(TreeCacheEventConnLost, nil)
+	case curator.CONNECTED:
+		tc.root.wasCreated()
+	case curator.RECONNECTED:
+		if err := tc.root.wasReconnected(); err == nil {
+			tc.publishEvent(TreeCacheEventConnReconnected, nil)
+		}
+	}
+}
+
+// publishEvent publish an event with given type and data to all listeners.
+func (tc *TreeCache) publishEvent(tp TreeCacheEventType, data *ChildData) {
+	if tc.state.Value() != curator.STOPPED {
+		evt := TreeCacheEvent{Type: tp, Data: data}
+		tc.logger.Debugf("publishEvent: %v", evt)
+		tc.callListeners(evt)
+	}
+}
```

Comment:
Maybe we should call the user listeners in a background goroutine? It is the default behavior of Apache Curator and it prevents listeners from blocking each other.